Set `-DLIBOMP_FORTRAN_MODULES=ON` on windows

h-vetinari opened the pull request:
If we want to build OpenBLAS with OpenMP support with flang (c.f. https://github.com/conda-forge/openblas-feedstock/pull/115), we'll need to package the Fortran bits of OpenMP, otherwise we run into:
```
[...]
[3646/19184] Building Fortran object CMakeFiles\LAPACK_OVERRIDES.dir\lapack-netlib\SRC\ssytrd_2stage.f.obj
[3647/19184] Building Fortran object CMakeFiles\LAPACK_OVERRIDES.dir\lapack-netlib\SRC\ssytrd_sb2st.F.obj
FAILED: CMakeFiles/LAPACK_OVERRIDES.dir/lapack-netlib/SRC/ssytrd_sb2st.F.obj 
%BUILD_PREFIX%\Library\bin\flang.exe -I%SRC_DIR%\lapack-netlib\SRC -I%SRC_DIR%\lapack-netlib\LAPACKE\include -fopenmp -fopenmp -ffixed-line-length-72 -o CMakeFiles\LAPACK_OVERRIDES.dir\lapack-netlib\SRC\ssytrd_sb2st.F.obj -c CMakeFiles\LAPACK_OVERRIDES.dir\lapack-netlib\SRC\ssytrd_sb2st.F-pp.f
error: Semantic errors in CMakeFiles\LAPACK_OVERRIDES.dir\lapack-netlib\SRC\ssytrd_sb2st.F-pp.f
D:\\bld\\openblas_1719536014536\\work\\lapack-netlib\\SRC\\ssytrd_sb2st.F:237:11: error: Cannot read module file for module 'omp_lib': Source file 'omp_lib.mod' was not found
        use omp_lib
            ^^^^^^^
```

Flang 18 is going to be [unusable](https://github.com/llvm/llvm-project/issues/86463) to build OpenBLAS, but I'd like to open this for discussion/review already, so we can figure this out before LLVM 19 releases.

PS. This is rebased from the `dev` branch, with which I was able to get CI green in https://github.com/conda-forge/openblas-feedstock/pull/115.

h-vetinari commented:
This is quite curious. We cannot build the fortran bits of OpenMP 18 with flang 18, but it works if _either_ of the versions is bumped to 19. In this PR we bump the flang version, on the `dev` branch it's the other way 'round.

In any case, even if we don't merge this for OpenMP 18, I want to enable this for OpenMP 19, so I'd appreciate feedback @conda-forge/openmp

h-vetinari commented:
Gentle ping on this @conda-forge/openmp

isuruf commented:
Can we move the `.mod` file to a different output? The cycles are also there in linux if we switch `_openmp_mutex`.

isuruf commented on recipe/meta.yaml:
```diff
@@ -21,26 +21,32 @@ source:
         - _openmp_mutex
+      ignore_run_exports_from:
+        - {{ compiler("cxx") }}
+        - {{ compiler("fortran") }}
```
The `.mod` files are specific to the gfortran version (maybe not the major version, but the soversion. not sure), so it's not possible to remove these.

conda-forge-admin commented:
Hi! This is the friendly automated conda-forge-linting service.

I wanted to let you know that I linted all conda-recipes in your PR (```recipe/meta.yaml```) and found some lint.

Here's what I've got...


For **recipe/meta.yaml**:

 * The build section contained an unexpected subsection name. build is not a valid subsection name.
 * The build section contained an unexpected subsection name. host is not a valid subsection name.

conda-forge-admin commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipe/meta.yaml```) and found it was in an excellent condition.

h-vetinari commented:
> Can we move the `.mod` file to a different output?

@isuruf, I've rebased this PR on #144 and added the output you wanted - I think I've incorporated all the feedback so far on this here. I still want to clean up the commit history better, but the _content_ of the changes should be OK now. As such, I'd ask you to please review this at your earliest convenience.

h-vetinari commented:
Gentle ping @isuruf

isuruf commented on recipe/meta.yaml:
```diff
@@ -71,6 +84,37 @@ outputs:
+      run:
+        - {{ pin_subpackage("llvm-openmp", exact=True) }}
+      run_constrained:
+        - llvm-openmp {{ version }}
```
```suggestion
```

isuruf commented on recipe/meta.yaml:
```diff
@@ -71,6 +84,37 @@ outputs:
+        - {{ pin_subpackage("llvm-openmp", exact=True) }}
+      run:
+        - {{ pin_subpackage("llvm-openmp", exact=True) }}
+      run_constrained:
```
We need a run_constrained on `flang` here as the mod file made here might only work with the flang compiler version used.

h-vetinari commented:
Thanks for the review; should be fixed now.

isuruf commented on recipe/meta.yaml:
```diff
@@ -71,6 +85,37 @@ outputs:
+      run:
+        - {{ pin_subpackage("llvm-openmp", exact=True) }}
+      run_constrained:
+        - {{ compiler("fortran") }} =={{ fortran_compiler_version }}
```
```suggestion
        - {{ compiler("fortran") }}
```

isuruf replied:
version is implicit in the jinja

h-vetinari replied:
Ah true, that already contains the version...

h-vetinari commented:
Good now?

h-vetinari commented:
I think we're close here @isuruf, can you please help me wrap this up?
